lib/OA/Admin/PasswordRecovery.php

```diff
@@ -217,37 +217,66 @@ function checkAccess()
         $aUsers = $this->_dal->searchMatchingUsers($email);
 
-        $aEmails = array();
+        // Send a separate password reset link in an email for each
+        // of the users found that match the email address
         foreach ($aUsers as $u) {
-            $aEmails[$u['email_address']][] = $u;
-        }
-
-        $sent = 0;
-        foreach ($aEmails as $email => $aUsers) {
-            $text = '';
-            foreach ($aUsers as $u) {
-                $recoveryId = $this->_dal->generateRecoveryId($u['user_id']);
-
-                $header = $GLOBALS['strUser']." {$u['contact_name']}";
-                $text .= $header."\n".str_repeat('-', strlen($header))."\n";
-                $text .= $GLOBALS['strUsername'].": {$u['username']}\n";
-                $text .= $GLOBALS['strPwdRecResetLink'].": ";
-                $text .= Max::constructURL(MAX_URL_ADMIN, "password-recovery.php?id={$recoveryId}")."\n\n";
+            // Generate the password reset email subject
+            $emailSubject = sprintf($GLOBALS['strPwdRecEmailPwdRecovery'], PRODUCT_NAME);
+
+            // Set the name and email address that the password reset email
+            // is to be sent from. Ideally, this will be the administrative
+            // account of the Revive Adserver installation - but fall back
+            // to the user's own email address if required
+            if (!empty($aConf['email']['fromName'])) {
+                $emailFromName = $aConf['email']['fromName'];
+            } else {
+                $emailFromName = $email;
+            }
+            if (!empty($aConf['email']['fromAddress'])) {
+                // Use the administrative account details
+                $emailFromAddress = $aConf['email']['fromAddress'];
+            } else {
+                $emailFromAddress = $email;
             }
 
-            // Hack
-            $aConf['email']['admin_name'] = $aPref['admin_fullname'];
-            $aConf['email']['admin']      = $aPref['admin_email'];
+            // Generate the password reset URL for this user
+            $recoveryId = $this->_dal->generateRecoveryId($u['user_id']);
+            $recoveryUrl = Max::constructURL(MAX_URL_ADMIN, "password-recovery.php?id={$recoveryId}");
+
+            // Load the appropriate language details for the email recipient
+            Language_Loader::load('default', $u['language']);
+
+            // Generate the body of the password reset email for this user
+            $emailBody = $GLOBALS['strPwdRecEmailBody'];
+            $emailBody = str_replace('{name}', $u['contact_name'], $emailBody);
+            $emailBody = str_replace('{username}', $u['username'], $emailBody);
+            $emailBody = str_replace('{reset_link}', $recoveryUrl, $emailBody);
+            if (!empty($aConf['email']['fromName']) && !empty($aConf['email']['fromAddress'])) {
+                $adminSignature = "Sincerely,\n\n{$aConf['email']['fromName']}\n{$aConf['email']['fromAddress']}";
```

Comment:
Signatures should probably be translatable too.

Reply:
Yes! Done.